Origin: 5e14da97f16e421d084a9e735be21b1025150f0c (pixman.h)
 2f876cf86718d3dd9b3b04ae9552530edafe58a1 (trap-crasher.c)
Description: pixman_trapezoid_valid(): Fix underflow when bottom is close to
 MIN_INT
 If t->bottom is close to MIN_INT (probably invalid value), subtracting top can
 lead to underflow which causes crashes. Attached patch will fix the issue.
Bug: https://bugs.freedesktop.org/show_bug.cgi?id=67484
Ubuntu-Bug: https://launchpad.net/bugs/1197921

Index: pixman-0.24.4/pixman/pixman.h
===================================================================
--- pixman-0.24.4.orig/pixman/pixman.h	2013-12-03 12:15:58.000000000 -0600
+++ pixman-0.24.4/pixman/pixman.h	2013-12-03 12:15:58.000000000 -0600
@@ -909,7 +909,7 @@
 #define pixman_trapezoid_valid(t)				   \
     ((t)->left.p1.y != (t)->left.p2.y &&			   \
      (t)->right.p1.y != (t)->right.p2.y &&			   \
-     (int) ((t)->bottom - (t)->top) > 0)
+     ((t)->bottom > (t)->top))
 
 struct pixman_span_fix
 {
Index: pixman-0.24.4/test/trap-crasher.c
===================================================================
--- pixman-0.24.4.orig/test/trap-crasher.c	2013-12-03 12:15:58.000000000 -0600
+++ pixman-0.24.4/test/trap-crasher.c	2013-12-03 12:15:58.000000000 -0600
@@ -5,13 +5,25 @@
 main()
 {
     pixman_image_t *dst;
-    pixman_trapezoid_t traps[1] = {
+    pixman_trapezoid_t traps[] = {
 	{
 	    2147483646,
 	    2147483647,
 	    {
 		{ 0, 0 },
 		{ 0, 2147483647 }
+	    },
+	    {
+		{ 65536, 0 },
+		{ 0, 2147483647 }
+	    }
+	},
+	{
+	    32768,
+	    - 2147483647,
+	    {
+		{ 0, 0 },
+		{ 0, 2147483647 }
 	    },
 	    {
 		{ 65536, 0 },
